<a href="https://colab.research.google.com/github/AdelinaKhairullinaUfa/-/blob/main/%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B52.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [51]:
import pandas as pd
df = pd.read_csv('СМАД Тема 3 ДЗ 2.csv')
df

age workclass  fnlwgt     education  education.num      marital.status  \
0       90         ?   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66         ?  186061  Some-college             10             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
0                      ?  Not-in-family  White  Female             0   
1        Exec-managerial  Not-in-family  White  Female             0   
2                      ?      Unmarried  Black  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country income  
0              4356              40  United-States  <=50K  
1              4356              18  United-States  <=50K  
2              4356              40  United-States  <=50K  
3              3900              40  United-States  <=50K  
4              3900              40  United-States  <=50K  
...             ...             ...            ...    ...  
32556             0              40  United-States  <=50K  
32557             0              38  United-States  <=50K  
32558             0              40  United-States   >50K  
32559             0              40  United-States  <=50K  
32560             0              20  United-States  <=50K  

[32561 rows x 15 columns]

In [52]:
import numpy as np
missing_values = df.isin(['?', np.nan]).sum()
print("Пропущенные значения:")
print(missing_values)


Пропущенные значения:
age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64


In [53]:
# Замена пропущенных значений "?" на NaN
df.replace('?', np.nan, inplace=True)

# Удаление строк с пропущенными значениями
df.dropna(inplace=True)

# Преобразование категориальных переменных с помощью One-Hot Encoding
df = pd.get_dummies(df, columns=['income'])
df

age workclass  fnlwgt     education  education.num      marital.status  \
1       82   Private  132870       HS-grad              9             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
5       34   Private  216864       HS-grad              9            Divorced   
6       38   Private  150601          10th              6           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
1        Exec-managerial  Not-in-family  White  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
5          Other-service      Unmarried  White  Female             0   
6           Adm-clerical      Unmarried  White    Male             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country  income_<=50K  income_>50K  
1              4356              18  United-States          True        False  
3              3900              40  United-States          True        False  
4              3900              40  United-States          True        False  
5              3770              45  United-States          True        False  
6              3770              40  United-States          True        False  
...             ...             ...            ...           ...          ...  
32556             0              40  United-States          True        False  
32557             0              38  United-States          True        False  
32558             0              40  United-States         False         True  
32559             0              40  United-States          True        False  
32560             0              20  United-States          True        False  

[30162 rows x 16 columns]

In [54]:
print(df.columns)

Index(['age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country',
       'income_<=50K', 'income_>50K'],
      dtype='object')


In [55]:
education_grouped_income = df.groupby('education').agg({'income_<=50K': 'mean', 'income_>50K': 'mean'})
print("Средний доход для каждой группы образования:")
print(education_grouped_income)

# Изучим распределение целевой переменной "income"
income_distribution = df['income_<=50K'].value_counts(normalize=True)
print("\nРаспределение целевой переменной income:")
print(income_distribution)

# Рассчитаем статистические показатели для каждой колонки
column_statistics = df.describe()
print("\nСтатистика для каждой колонки:")
print(column_statistics)

# Найдем группу с максимальным и минимальным доходом
max_income_group = education_grouped_income['income_>50K'].idxmax()
min_income_group = education_grouped_income['income_>50K'].idxmin()

max_income_value = education_grouped_income.loc[max_income_group, 'income_>50K']
min_income_value = education_grouped_income.loc[min_income_group, 'income_>50K']

print("\nГруппа с максимальным доходом:", max_income_group, "(", max_income_value, ")")
print("Группа с минимальным доходом:", min_income_group, "(", min_income_value, ")")
df

Средний доход для каждой группы образования:
              income_<=50K  income_>50K
education                              
10th              0.928049     0.071951
11th              0.943702     0.056298
12th              0.923077     0.076923
1st-4th           0.960265     0.039735
5th-6th           0.958333     0.041667
7th-8th           0.937163     0.062837
9th               0.945055     0.054945
Assoc-acdm        0.746032     0.253968
Assoc-voc         0.736802     0.263198
Bachelors         0.578509     0.421491
Doctorate         0.253333     0.746667
HS-grad           0.835671     0.164329
Masters           0.435771     0.564229
Preschool         1.000000     0.000000
Prof-school       0.250923     0.749077
Some-college      0.799940     0.200060

Распределение целевой переменной income:
income_<=50K
True     0.751078
False    0.248922
Name: proportion, dtype: float64

Статистика для каждой колонки:
                age        fnlwgt  education.num  capital.gain  capital.loss  \

age workclass  fnlwgt     education  education.num      marital.status  \
1       82   Private  132870       HS-grad              9             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
5       34   Private  216864       HS-grad              9            Divorced   
6       38   Private  150601          10th              6           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
1        Exec-managerial  Not-in-family  White  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
5          Other-service      Unmarried  White  Female             0   
6           Adm-clerical      Unmarried  White    Male             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country  income_<=50K  income_>50K  
1              4356              18  United-States          True        False  
3              3900              40  United-States          True        False  
4              3900              40  United-States          True        False  
5              3770              45  United-States          True        False  
6              3770              40  United-States          True        False  
...             ...             ...            ...           ...          ...  
32556             0              40  United-States          True        False  
32557             0              38  United-States          True        False  
32558             0              40  United-States         False         True  
32559             0              40  United-States          True        False  
32560             0              20  United-States          True        False  

[30162 rows x 16 columns]

In [64]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error

# Преобразование категориальных переменных методом One-Hot Encoding и заполнение пропущенных значений
X_encoded = pd.get_dummies(X)
X_encoded.fillna(0, inplace=True)

X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Обучение модели линейной регрессии
model = LinearRegression()
model.fit(X_train, y_train)


y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print("Метрика R^2:", round(r2, 2))
print("Метрика MSE:", round(mse, 2))



Метрика R^2: 0.35
Метрика MSE: 0.12



 R^2 (коэффициента детерминации) равно приблизительно

35% дисперсии зависимой переменной (доход) объясняется использованными независимыми переменными (признаками). Это не очень высокий уровень объясняемой дисперсии, что может говорить о том, что модель не очень хорошо подходит для предсказания дохода.

Среднеквадратичная ошибка (MSE) равна приблизительно
0.12. MSE представляет собой среднее значение квадратов разностей между фактическими и предсказанными значениями. Чем ниже значение MSE, тем лучше. В данном случае, значение MSE довольно низкое, что указывает на то, что модель имеет неплохую точность в предсказании дохода.

Таким образом, можно сделать вывод, что хотя модель может предсказывать доход с некоторой точностью (как указывает низкое значение MSE), она объясняет только небольшую часть дисперсии зависимой переменной